blockstack/lib/nameset/virtualchain_hooks.py

```diff
@@ -565,13 +565,44 @@ def db_save( block_height, consensus_hash, ops_hash, accepted_ops, virtualchain_
             log.error("FATAL: failed to update subdomains db at {}".format(block_height))
             os.abort()
 
+
+        # TODO: create a new RPC endpoint to expose the state changes below, for testing when namespace signal threshold reached, and when import block is reached
+
+        # namespace to watch for miner signals to perform v2 upgrade
+        v2_upgrade_signal_namespace = 'miner'
+        # threshold of required ID registrations
+        v2_upgrade_signal_required_registrations = 20
+        # number of blocks to wait after signal threshold is reached until block import and chainstate dump 
+        v2_upgrade_signal_import_threshold = 500
+        if BLOCKSTACK_TEST:
+            v2_upgrade_signal_import_threshold = 10
+
+        # check if the ID threshold block has already been reached
+        threshold_block_id = db_state.get_v2_upgrade_threshold_block()
+        if threshold_block_id is not None:
+            import_block = db_state.get_v2_import_block_reached()
+            if import_block is None:
+                # check if the required number of blocks have been mined since threshold crossed
+                if (block_height - threshold_block_id == v2_upgrade_signal_import_threshold):
+                    # emit parsable log entry and export datafile
+                    log.warn('[v2-upgrade] import threshold reached at block: {}'.format(block_height))
+                    db_state.set_v2_import_block_reached(block_height)
+                    db_state.perform_v2_upgrade_datafile_export(consensus_hash)
```

Comment:
Do you have an idea yet about how long this will take?  It's okay if it completely stalls the node, but we'll want to make sure that the export gets created in time for v2 to launch (which I think is 144 blocks after the threshold is met?)

Reply:
Yes, the follow-up PR addresses this: https://github.com/blockstack/stacks-blockchain/pull/1974